LAMMPS parser: system parsing

Bernadette-Mohr opened the pull request:
This PR addresses migrating `parse_system` of the LAMMPS parser from the old version in `atomisticparsers` to the new `ModelSystem`.

**TODO**
- [x] Update LAMMPS parser to new schema -- parse system only, without molecule hierarchy
- [x] Transfer tests from old parser, comment out non-system tests for now
- [x] Transfer Leo's changes to old parser - Mostly commented out and untested
- [x] Test parsing of CG systems -- should work using ParticleState()

ladinesa committed:
Initial lammps parser

ladinesa committed:
Rebasing develop

Bernadette-Mohr committed:
removed git dependency from pyproject.toml

Bernadette-Mohr committed:
Initial stages of migrating the LAMMPS parser

Bernadette-Mohr committed:
Initial stages of migrating the LAMMPS parser

Bernadette-Mohr committed:
Saving progress

Bernadette-Mohr committed:
Saving progress

Bernadette-Mohr committed:
WIP commit

Bernadette-Mohr committed:
WIP commit

Bernadette-Mohr committed:
Rebasing develop to update

Bernadette-Mohr committed:
Updating System and Methods

Bernadette-Mohr committed:
WIP save status quo

Bernadette-Mohr committed:
Rebasing develop for updates

Bernadette-Mohr committed:
version incompatibility between MDParser and ModelSystem

Bernadette-Mohr committed:
Rebasing to develop for updates

Bernadette-Mohr committed:
Updating some mdparserutils functionalities

Bernadette-Mohr committed:
Updated MDParserutils.parse_trajectory_step according to new ModelSystem. Temporarily uncommented ForceContribution and EnergyContribution due to ongoing changes in nomad-simulations

Bernadette-Mohr committed:
Rebasing to develop for updates

Bernadette-Mohr committed:
Rebasing to develop for updates

Bernadette-Mohr committed:
Start migrating of tests

Bernadette-Mohr committed:
Moved test_lammpsparser to correct directory

Bernadette-Mohr committed:
Moved test_lammpsparser to correct directory

Bernadette-Mohr committed:
delete lammps parser test from wrong position

Bernadette-Mohr committed:
delete lammps parser test from wrong position

Bernadette-Mohr committed:
Renamed lammps tests accroding to conventions, updated tests for migrated system parsing

Bernadette-Mohr committed:
Renamed lammps tests accroding to conventions, updated tests for migrated system parsing

Bernadette-Mohr committed:
Migrated parse_system except system hierarchy part. Need more information/example data to fully integrate Leo's fixes

Bernadette-Mohr committed:
Migrated parse_system except system hierarchy part. Need more information/example data to fully integrate Leo's fixes

Bernadette-Mohr committed:
Merge branch 'develop' into lammps-parser

Bernadette-Mohr committed:
Updated inheritance from ArchiveWriter to current version of MDParser

Bernadette-Mohr committed:
Updated inheritance from ArchiveWriter to current version of MDParser

JFRudzinski commented on pyproject.toml:
```diff
@@ -33,7 +33,7 @@ maintainers = [
 dependencies = [
     "nomad-lab>=1.3.16.dev100",
-    "nomad-simulations@git+https://github.com/nomad-coe/nomad-simulations.git@develop",
+    "nomad-simulations",  # @git+https://github.com/nomad-coe/nomad-simulations.git@develop",
```
you should not have to change these dependencies when you work in the distro dev, don't forget to change it back in the end

Bernadette-Mohr replied:
I unfortunately do have to do that, which doesn't seem to be too surprising: 
https://discord.com/channels/1201445470485106719/1383005977586172004

I'll make a habit of changing it back before I request a review.

Currently, I'm not seeing this behavior anymore!

JFRudzinski commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1453 @@
+                value = value * units.get(unit, 1)
+            return value
+
+        def get_composition(children_names):
```
fyi - composition_formula will now be set by the Simulation normalizer in nomad-simulations, so you do not need to set this in the parser

Bernadette-Mohr replied:
Thank you for the information, I removed the function from the parser.

JFRudzinski commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1453 @@
+
+        units = self.log_parser.units
+
+        def apply_unit(value, unit):
```
This probably should be a class method right? I imagine it will be used by other parsing functions

Bernadette-Mohr replied:
It is not used by any other function as far as I can see now. I changed it to a class method for now, and will be cleaning up the parser structure in the end to address issues like this.

JFRudzinski reviewed (commented):
looking good so far, just a few minor comments upon browsing

Bernadette-Mohr committed:
Minor updates to accomodate changes brought by merge with develop

Bernadette-Mohr committed:
Rebasing develop for updates

Bernadette-Mohr committed:
Started minor cleanup of the parser structure

Bernadette-Mohr committed:
Started minor cleanup of the parser structure

Bernadette-Mohr committed:
Changed get_pbc_cell to class method of TrajParser, added unit test for cell types

Bernadette-Mohr committed:
Changed get_pbc_cell to class method of TrajParser, added unit test for cell types

Bernadette-Mohr committed:
Updated LB's latest edits

Bernadette-Mohr committed:
Updated LB's latest edits

Bernadette-Mohr committed:
Merge branch 'develop' into lammps-parser

Bernadette-Mohr committed:
minor edits

Bernadette-Mohr committed:
minor edits

Bernadette-Mohr committed:
Changed dummy label for non-atomic particles to 'CGX' to bypass ASE check

Bernadette-Mohr committed:
Changed dummy label for non-atomic particles to 'CGX' to bypass ASE check

ndaelman-hu committed:
Fix import bug: path and classname (#44)

Fixes an import bug by correcting the module path and class name for the DFT workflow in the Wannier90 parser.

    Updates import statement to use the correct module path (dmft instead of beyond_dft)
    Changes class name from DFTTBWorkflow to DFTTBDMFTWorkflow to match the corrected import

Co-authored-by: ndaelman <ndaelman@physik.hu-berlin.de>

JFRudzinski commented:
@ladinesa Just a reminder in case you lost track of this one. Could you give this code a review whenever you have a chance (no rush). Thanks!

ladinesa commented:
sorry a bit confused, I implemented an initial parser for lammps and it was extended and merged. This pr aims to extend  and allign with the changes in the model system  defs. But seeing the changes, it seems that it is introducing a new parser. Can you enlighten me a bit?

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+from nomad_simulations.schema_packages.model_method import ModelMethod
+from nomad_simulations.schema_packages.model_system import Cell, ModelSystem
+
+from simulationworkflowschema import (
```
I suggest you use the workflow defs in nomad-simulations

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+    GeometryOptimizationMethod,
+    GeometryOptimizationResults,
+)
+from simulationworkflowschema.molecular_dynamics import (
```
oh, this one ok, maybe use the old schema but please create an issue to change this to the new schema

Bernadette-Mohr replied:
@JFRudzinski is currently working on migrating `molecular_dynamics`, I will update the dependencie accordingly.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+)
+from structlog.stdlib import BoundLogger
+
+re_float = r'[-+]?\d+\.*\d*(?:[Ee][-+]\d+)?'
```
maybe put this in utils? or make use all caps `RE_FLOAT`

Bernadette-Mohr replied:
I've moved those to `utils.constants_definitions` for now. If acceptable, I will update their use in other parsers.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+re_n = r'[\n\r]'
+
+
+def get_unit(units_type, property_type=None, dimension=3):
```
specify types for args and return value

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+
+
+def get_unit(units_type, property_type=None, dimension=3):
+    mole = 6.022140857e23
```
put in utils?

Bernadette-Mohr replied:
I have moved the definition to `utils.constants_definitions` as well.

I am wondering why we aren't using `ureg.mole` here instead of carrying an additional definition of Avogadro's number through all parsers I am familiar with? Will test if using any of the existing `pint` definitions would work.

JFRudzinski replied:
ureg.mole is a unit, this number is a constant, IMO, badly named, it should be NA (i.e., Avogadro's number)

Bernadette-Mohr replied:
True, ureg.avogadro_constant or ureg.avogadro_number are probably closer to what we need, but they are already defined as well.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        self._chemical_symbols = None
+        super().__init__(None)
+
+    def get_pbc_cell(self, val) -> tuple[list, np.ndarray]:
```
specify type for arg

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+
+    def init_quantities(self):
+        def get_atoms_info(val):
+            val = val.split('\n')
```
use splitlines()

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+    def n_frames(self) -> int:
+        return len(self.get('atoms_info', []))
+
+    @property
```
why make this as property when there is no update logic? maybe simply have self.masses or an update logic like `if self._masses is None: ....`

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        return self._masses
+
+    # TODO: handle non-atomistic representations
+    @masses.setter
```
ah you have a setter, forget my other comment

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+
+    # TODO: handle non-atomistic representations
+    @masses.setter
+    def masses(self, val):
```
type for arg

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+    # TODO: handle non-atomistic representations
+    @masses.setter
+    def masses(self, val):
+        self._masses = val
```
shouldnt this be ```
if val is None:
    return```

Bernadette-Mohr replied:
Absolutely! It's changed now.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        if self._masses is None:
+            return
+
+        self._masses = val
```
put a sanity check on val before setting, in line 301 you are assuming it to be at least a 2d array

Bernadette-Mohr replied:
I've added a check for `val` being a numpy array of the appropriate dimension. Please suggest a more elegant solution if mine isn't optimal!

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+class TrajParser(TextParser):
+    def __init__(self):
+        self._masses = None
+        self._reference_masses = dict(
```
I suggest you make masses and chem symbols as global vars or include it in utils

Bernadette-Mohr replied:
`REFERENCE_MASSES` and `CHEMICAL_SYMBOLS` (both internal definitions adopted from ASE) are defined as constants in `utils.constants_definitions` now.

Also, I separated the derivation of system-applicable chemical symbols from the masses setter.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+                    'symbols'
+                ][symbol_idx]
+
+    def get_atom_labels(self, idx):
```
type for arg and return val

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        if atoms_info is None:
+            return
+
+        atoms_id = atoms_info[idx].get('id')
```
more robust: `atoms_info.get(idx, {}).get('id')`

Bernadette-Mohr replied:
I also changed the syntax for `atoms_type` accordingly.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+
+        atoms_id = atoms_info[idx].get('id')
+        default = ['CGX' for _ in atoms_id] if atoms_id is not None else None
+        atoms_type = atoms_info[idx].get('type')
```
maybe define`id_info = atoms_info.get(idx)` at the top

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        atoms_type = atoms_info[idx].get('type')
+        if atoms_type is None:
+            return default
+        if self._chemical_symbols is None:
```
shouldn't this be at the start? oh you determine default, sorry

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+            atom_labels = [self._chemical_symbols[atype] for atype in atoms_type]
+        except Exception:
+            self.logger.error('Error resolving atom labels.')
+            return
```
maybe the return val is [] or simply set atom_labels = [] actually better you do not need the try as there seems to be exceptions to be had just do `self._chemical_symbols.get(atype)` then filter return [label for label in atom_labels if label is not None]

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+
+        return atom_labels
+
+    def get_positions(self, idx):
```
arg and return type

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        atoms_info = atoms_info[idx]
+
+        cell = self.get('pbc_cell')
+        cell = None if cell is None else cell[idx][1]
```
shouldnt you have `if is cell is None:
return` here?

ladinesa replied:
ah you have the l349 case, put this as the first case then all that requires cell in elif and return if cell is None

ladinesa replied:
just in general I discourage using the return statement right away as it makes extending the code a little trickier, set instead the return value to something and have only one return statement at the end

Bernadette-Mohr replied:
I've changed the flow a bit, to improve readability and minimize redundant checks for the availiability of `cell`. There's only one final `return` at the end now.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+
+        return positions
+
+    def get_velocities(self, idx):
```
type for arg and return val

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        return positions
+
+    def get_velocities(self, idx):
+        atoms_info = self.get('atoms_info')
```
this is becoming so repetitive I suggest having a common function for this.

Bernadette-Mohr replied:
I created helper functions for accessing `atoms_info` and coordinates, and restructured `get_positions`, `get_velocities`, and `get_forces`.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        if 'vx' not in atoms_info or 'vy' not in atoms_info or 'vz' not in atoms_info:
+            return
+
+        return np.array([atoms_info['vx'], atoms_info['vy'], atoms_info['vz']]).T
```
np.transpose([atoms_info.get(key) for key in (vs, vy, vz)])

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+
+        return np.array([atoms_info['vx'], atoms_info['vy'], atoms_info['vz']]).T
+
+    def get_forces(self, idx):
```
type for arg and return val

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        atoms_info = atoms_info[idx]
+        if 'fx' not in atoms_info or 'fy' not in atoms_info or 'fz' not in atoms_info:
+            return
+        return np.array([atoms_info['fx'], atoms_info['fy'], atoms_info['fz']]).T
```
I suggest having a common function for get_forces and get_velocities e.g. get_atoms_array? that take in the key either f or v.

Bernadette-Mohr replied:
`_get_frame_atoms_info` and `_extract_vector_components` are now used by `get_positions`, `get_velocities`, and `get_forces`.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+            return
+        return np.array([atoms_info['fx'], atoms_info['fy'], atoms_info['fz']]).T
+
+    def get_lattice_vectors(self, idx):
```
arg and return val type

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        pbc_cell = self.get('pbc_cell')
+        if pbc_cell is None:
+            return
+        return pbc_cell[idx][1]
```
is pbc_cell a dict? return self.get('pbc_cell', {}).get(idx, [None])[0]

Bernadette-Mohr replied:
`pbc_cell` is a list of tuples, so no `get`. I have increased the defensiveness of the new `_get_cell_component` function.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+            return
+        return pbc_cell[idx][1]
+
+    def get_pbc(self, idx):
```
have a common function for get_lattice_vectors and get_pbc taking in 0 or 1 as arg.

Bernadette-Mohr replied:
See comment above, there is `_get_cell_component` now.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+            return
+        return pbc_cell[idx][0]
+
+    def get_n_atoms(self, idx):
```
type for arg and return val

ladinesa replied:
have a common function for get_n_atoms and get_step taking in key and idx

Bernadette-Mohr replied:
Both functions use `_get_trajectory_info` now.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+
+
+class XYZTrajParser(TrajParser):
+    def __init__(self):
```
you do not need this if there is nothing to init

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+
+    def init_quantities(self):
+        def get_atoms_info(val_in):
+            val = [v.split('#')[0].split() for v in val_in.strip().split('\n')]
```
use splitlines()

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+            val = np.transpose(np.array([v for v in val if len(v) == 4], dtype=float))
+            # val[0] is the atomic number
+            val[0] = [list(set(val[0])).index(v) + 1 for v in val[0]]
+            return dict(type=val[0], x=val[1], y=val[2], z=val[3])
```
more pythonic return {key: val[n] for n, key in enumerate(['type', 'x', 'y', 'z'])}, but as it stands i'm fine with it.

Bernadette-Mohr replied:
I find your suggestion more readable, thank you!

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+
+
+class TrajParsers:
+    def __init__(self, parsers):
```
type for arg

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        for parser in parsers:
+            parser.parse()
+
+    def __getitem__(self, index):
```
type for arg and return val

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        if self._parsers:
+            return self._parsers[index]
+
+    def eval(self, key, *args, **kwargs):
```
- type for arg and return val

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+            for header in self._headers
+        ]
+
+        def get_section_value(val):
```
type for arg and return al

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        super().__init__(None)
+
+    def init_quantities(self):
+        def str_op(val):
```
type for arg and return val

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+    def init_quantities(self):
+        def str_op(val):
+            val = val.split('#')[0]
+            val = val.replace('&\n', ' ').split()
```
you also have \r as escape char?

Bernadette-Mohr replied:
Absolutely. Using the already existing constant `RE_N` for handling different escape chars.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        self._quantities = [
+            Quantity(
+                name,
+                r'\n\s*%s\s+(?!.*\$\{)([${}\w\. \/\#\-]+)(\&\n[\w\. \/\#\-]*)*' % name,
```
use f-string format

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+            )
+        )
+
+        def str_to_thermo(val):
```
arg and return val type

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+            res = {}
+            if val.count('Step') > 1:
+                val = (
+                    val.replace('--', '').replace('=', '').replace('(sec)', '').split()
```
you can do a regex sub instead

Bernadette-Mohr replied:
I've made the change, still need to test it.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+                        res[val[i]].append(float(val[i + 1]))
+
+            else:
+                val = val.split('\n')
```
splitlines

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+            self._units = get_unit(units_type)
+        return self._units
+
+    def get_thermodynamic_data(self):
```
return val type

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+                data[key] = val
+        return data
+
+    def get_traj_files(self):
```
return val type

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+            traj_files = [
+                f
+                for f in traj_files
+                if f.endswith('trj') or f.endswith('xyz')
```
use insead if f.split('.')[-1] in supported_traj files

ladinesa replied:
we have utils.general.search_files, you should use this

Bernadette-Mohr replied:
I am not sure why we cover the case of "no dump command in the log file" when identifying trajectory files? Afaik, without an explicit dump command, Lammps wouldn't ouput a trajectory in the first place.
But since there probably is a good reason for this logic being here, I extended the scanning capablility by `find_best_matching_file` to handle the fact that LAMPPS seems to enforce no file naming defaults or rules whatsoever.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+
+        return [os.path.join(self.maindir, f) for f in traj_files]
+
+    def get_data_files(self):
```
type for return val

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        return [os.path.join(self.maindir, f) for f in traj_files]
+
+    def get_data_files(self):
+        def check_file_header(file_path, regex_pattern):
```
type for arg and return val

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        #     except Exception:
+        #         pass
+        if read_data is None or 'CPU' in read_data:
+            self.logger.warning('Data file not specified in directory, will scan.')
```
define this warning  in init, oh different from Traj file

Bernadette-Mohr replied:
Do you mean it is ok to define this warning here, since it is for Lammps data files specifically, or would you still prefer both data- and trajectory file warning messages definied in the constructor?

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        #         pass
+        if read_data is None or 'CPU' in read_data:
+            self.logger.warning('Data file not specified in directory, will scan.')
+            data_files = os.listdir(self.maindir)
```
use search_files util

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+
+        return [os.path.join(self.maindir, f) for f in data_files]
+
+    def get_pbc(self):
```
return val type

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        pbc = self.get('boundary', ['p', 'p', 'p'])
+        return [v == 'p' for v in pbc]
+
+    def get_sampling_method(self):
```
retun val type

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        )
+        return sampling_method, fix_style
+
+    def get_thermostat_settings(self):
```
return val type

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        return sampling_method, fix_style
+
+    def get_thermostat_settings(self):
+        fix = self.get('fix', [None])[0]
```
instead of default [None] can be instead [[]] so you do not need l916-917

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+    def get_interactions(self):
+        styles_coeffs = []
+        for interaction in self._interactions:
+            styles = self.get('%s_style' % interaction, None)
```
use f-format

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        self.log_parser = LogParser()
+        self.units = self.log_parser.units
+        self.aux_log_parser = LogParser()
+        self._md_parser = MDParser()
```
be consistent if all parsers ought to be private

ladinesa replied:
i am confused, LammpsAchiveWriter is an MDParser but you have _md_parser

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        )
+        self.data_parser = DataParser()
+
+    def apply_unit(self, value, unit):
```
arg and return val type

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+            value = value * self.units.get(unit, 1)
+        return value
+
+    def parse_method(self, simulation):
```
why need to pass simulation?

ladinesa replied:
it seems nothing is set in this function, can you describe what is really going on here.

Bernadette-Mohr replied:
There's nothing set in this function yet, I prioritized porting `parse_system` to the new schema for testing. I'll get to `parse_method` as soon as I'm finished with the system.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+    #            )
+
+    def parse_system(self, simulation):
+        # sec_run = self.archive.run[-1]
```
remove

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+                'velocities': velocities,
+                'bond_list': self._bond_list if self._bond_list else None,
+            }
+            self._md_parser.parse_trajectory_step(particles_dict, simulation)
```
just do self.parser_trajectory_step here

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+    #                 neighmodify[index + 1]
+    #            )
+
+    def parse_system(self, simulation):
```
this is just too long, organize the contents and split them in chunks so it is easier to read

Bernadette-Mohr replied:
Agreed. I am currently reorganizing it on branch [`lammps-parser_system-hierarchy`](https://github.com/FAIRmat-NFDI/nomad-parser-plugins-simulation/pull/47) that holds the complete updated `parse_system`.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+                )
+            atoms_resnames = np.array(atoms_info.get('resnames', []))
+            moltypes = np.unique(atoms_moltypes)
+            # for i_moltype, moltype in enumerate(moltypes):
```
what is going on here?

Bernadette-Mohr replied:
Nothing further here on this PR, updating parsing the system hierarchy was decoupled and moved to [`lammps-parser_system-hierarchy`](https://github.com/FAIRmat-NFDI/nomad-parser-plugins-simulation/pull/47)

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+            #                 names_firstatom
+            #             )
+
+    def write_to_archive(self) -> None:
```
this is just too long, organize it into smaller chunks

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+        if len(traj_files) > 1:
+            self.logger.warning('Multiple traj files are specified')
+
+        parsers = []
```
you can have this as a separate function get_traj_parsers

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+
+        self.parse_system(self.archive.data)
+
+        # # include input controls from log file
```
either put TODO or remove

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,1485 @@
+    # ? Really needed for the LAMMPS parser?
+    # ? Would it make sense to handle a potential auxillary log file here,
+    # ? since there seems to be an issue with its handling in the old parser?
+    def is_mainfile(
```
remove, only need this for complex workflows

ladinesa commented on tests/parsers/test_lammps_parser.py:
```diff
@@ -0,0 +1,384 @@
+#     assert len(archive.run[0].system) == 4
+
+
+def test_traj_xyz(parser):
```
this should only use the XYZ parser not the main parser

Bernadette-Mohr replied:
While I'm at it changing the tests anyway, would you recommend replacing the test data files with fixtures/mock data/temporary files?

ladinesa commented on tests/parsers/test_lammps_parser.py:
```diff
@@ -0,0 +1,384 @@
+    assert sec_systems[13].positions[7][0].magnitude == pytest.approx(-8.00436e-10)
+
+
+def test_traj_dcd(parser):
```
same here

ladinesa commented on tests/parsers/test_lammps_parser.py:
```diff
@@ -0,0 +1,384 @@
+
+def test_unwrapped_pos(parser):
+    archive = EntryArchive()
+    parser.parse('tests/data/lammps/1_xyz_files/log.lammps', archive, LOGGER)
```
you only parse the file containing the pos and use only the relevant parser

ladinesa commented on tests/parsers/test_lammps_parser.py:
```diff
@@ -0,0 +1,384 @@
+
+LOGGER = get_logger(__name__)
+
+
```
organize your test in the same way you organize your classes in the parser e.g.
```
class TestXYYParser:
 ...

class LogParser:
...

class TestLammpsArchiveWriter:
...

```

and test individual functionalities.

Bernadette-Mohr replied:
Done, as far as the structure is clear so far. I'll improve on it as I go.

ladinesa commented:
Please split and organize the file parsers into several files, see the other parsers in this repo. I did not look the tests in greater detail but please organize the tests in the same way you organize your classes. Also remove all comments that were taken from the old parser. just put a TODO e.g. in tests

Bernadette-Mohr committed:
Started adding type hints to all functions, minor corrections

Bernadette-Mohr committed:
Started adding type hints to all functions, minor corrections

Bernadette-Mohr committed:
Finish rebasing to develop for updates

Bernadette-Mohr committed:
Re-added h5md parser entry point

Bernadette-Mohr committed:
Adding type hints to functions

Bernadette-Mohr committed:
Added type hints in parser, made tests more specific

Bernadette-Mohr committed:
Moved auxillary parsers to individual files, rearranged tests

Bernadette-Mohr committed:
reverted dependency setting in pyproject.toml

Bernadette-Mohr committed:
Removed print statements, changed workflow schema to nomad_simulations

Bernadette-Mohr committed:
Continued implementing PR comments

Bernadette-Mohr committed:
Continued code cleanup

Bernadette-Mohr committed:
Extended LogParser scanning routine - LAMMPS has no defined file naming convention

Bernadette-Mohr committed:
Finished updating auxilary parsers, started restructuring main parser

Bernadette-Mohr committed:
Removed leftover code from previous schema

Bernadette-Mohr committed:
Restructured write_to_archive

Bernadette-Mohr committed:
Added early exit in parse_trajectory_step if setting up of archive failed

Bernadette-Mohr committed:
Deleted unused imports, added some comments

JFRudzinski reviewed (commented):
@Bernadette-Mohr I suggest making a small TODO statement if there are any small uncertainties left from Alvin's extensive review, and then resolve those comments. If there are any comments that you cannot make a single-line TODO to double check later, then tag Alvin again so it is clear what he should check.

I think these will get naturally addressed as you fill the later sections of the parser code and if not you can ensure that they get tackled in a later PR where there is not so much noise from other issues.

When you are done with that and fix the formatting I will check and approve.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/file_parsers.py:
```diff
@@ -0,0 +1,771 @@
+
+                try:
+                    value.append(np.array(v, dtype=float))
+                # ? Should we append None here, instead of leaving value potentially an empty list?
```
you can put a return name, None in l277 instead of break

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/file_parsers.py:
```diff
@@ -0,0 +1,771 @@
+
+class LogParser(TextParser):
+    # TODO: add parsers for other LAMMPS-supported trajectory formats
+    SUPPORTED_TRAJ_EXTENSIONS = ['trj', 'xyz', 'dcd', 'lammpstrj']  # , 'h5', 'nc'
```
normally, only global vars are all caps, I suggest `_supported_traj_extensions` and `_data_file_patterns`

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/file_parsers.py:
```diff
@@ -0,0 +1,771 @@
+
+    def __init__(self) -> None:
+        self._commands = [
+            'angle_coeff',
```
these lists are becoming very long, do you think we will have to extend later. Can you please make a list of these extra long lists and maybe we can put them in a `config.json`

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/file_parsers.py:
```diff
@@ -0,0 +1,771 @@
+        if self._units is None:
+            units_type = self.get('units', ['lj'])[0]
+            self._units = get_unit(units_type)
+        return self._units
```
please add this so we reset units and potential other private vars when we set log file:
```python
def reset(self):
    super().reset()
    self._units = None
```

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/file_parsers.py:
```diff
@@ -0,0 +1,771 @@
+            traj_files = []
+            for ext in self.SUPPORTED_TRAJ_EXTENSIONS:
+                found_files = search_files(
+                    pattern=f'*{ext}*',
```
should be f`*.{ext}`?

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/file_parsers.py:
```diff
@@ -0,0 +1,771 @@
+                MIN_CHARS = 2
+                return {t for t in tokens if len(t) > MIN_CHARS}
+
+            def get_best_from_scores(
```
there is this [module](https://docs.python.org/3/library/difflib.html#difflib.get_close_matches) maybe you can use it to sort the files or [bisect](https://docs.python.org/3/library/bisect.html). But looks good already.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/file_parsers.py:
```diff
@@ -0,0 +1,771 @@
+            # ? Looking for the biggest file size/keywords (result/final/dry ...) could make more sense?
+            # Fallback: Use first file
+            if not matching_files:
+                self.logger.warning(
```
do not use formatted string

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/file_parsers.py:
```diff
@@ -0,0 +1,771 @@
+        res = dict()
+        if fix_style.lower() == 'nvt':
+            try:
+                res['target_T'] = float(fix[5]) * temp_unit
```
if possible, convert fix to list of floats so we do not see these float() conversions

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/file_parsers.py:
```diff
@@ -0,0 +1,771 @@
+                    style = styles[i][0]
+
+                else:
+                    coeff = self.get('%s_coeff' % interaction)
```
use f'' format

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,542 @@
+        if self.traj_parsers.eval('n_frames') is None:
+            return
+
+        sec_method = ModelMethod()
```
please remove l49-154 so we have a clean code, we add them later in a separate mr

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,542 @@
+                )
+            atoms_resnames = np.array(atoms_info.get('resnames', []))
+            moltypes = np.unique(atoms_moltypes)
+            # for i_moltype, moltype in enumerate(moltypes):
```
please remove all commented lines add them in a separate mr

Bernadette-Mohr committed:
Fixed Ruff complaints

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,542 @@
+    def _configure_parsers(self) -> None:
+        """Configure all parsers with loggers and basic settings."""
+
+        def _setup_auxiliary_log_parser(aux_log_name) -> None:
```
you do not need to define this function, simply put the code to the if statement below.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,542 @@
+        # Configure data parser
+        self._data_parser.logger = self.logger
+
+    def _parse_data_files(self) -> None:
```
is this really _parse_ or _set_?

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,542 @@
+        if data_files:
+            self._data_parser.mainfile = data_files[0]
+
+    def _create_formatted_parser(
```
this is not used outside the scope of _create_trajectory_parser put it there if linting permits

ladinesa replied:
or maybe not posible as you have long _create_custom_parser.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,542 @@
+        cases where traj files can share the same parser.
+        """
+
+        def _get_trajectory_file_type(traj_file: str, index: int) -> str:
```
you do not need to define this, you run into a confusing code with all the sub-sub-funcs

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,542 @@
+        # TODO: provide support for other file types
+        return traj_parser
+
+    def _create_trajectory_parser(
```
in general, if you can squeeze these in their relevant function defs please do so.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,542 @@
+
+        return traj_parser
+
+    def _create_fallback_parser(self, traj_file: str) -> TrajParser:
```
you do not need to define this just insert the code if possible.

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/trajectory_parsers.py:
```diff
@@ -0,0 +1,339 @@
+        self._masses = None
+        self._chemical_symbols = None
+        super().__init__(None)
+
```
insert 
```
def reset(self):
    super().reset()
    self._masses = None
    self._chemical_symbols = None
```

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/trajectory_parsers.py:
```diff
@@ -0,0 +1,339 @@
+            try:
+                val = np.asarray(val)
+            except (ValueError, TypeError) as e:
+                raise ValueError(f'Cannot convert masses to array: {e}')
```
do not raise error log and return None instead

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/trajectory_parsers.py:
```diff
@@ -0,0 +1,339 @@
+                raise ValueError(f'Cannot convert masses to array: {e}')
+        min_dim = 2
+        if val.ndim < min_dim:
+            raise ValueError(f'masses must be at least 2D, got {val.ndim}D array')
```
same

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/trajectory_parsers.py:
```diff
@@ -0,0 +1,339 @@
+            self._derive_chemical_symbols()
+        return self._chemical_symbols
+
+    def _derive_chemical_symbols(self) -> None:
```
you do not need to define this. simply insert the code in l114

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/trajectory_parsers.py:
```diff
@@ -0,0 +1,339 @@
+            symbol_idx = np.argmin(abs(REFERENCE_MASSES - masses[i][1]))
+            self._chemical_symbols[masses[i][0]] = CHEMICAL_SYMBOLS[symbol_idx]
+
+    def get_atom_labels(self, idx: int) -> list[str] | None:
```
i suggest you return empty list instead of None and look at the usage. inspect also other parsers for consistency

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/trajectory_parsers.py:
```diff
@@ -0,0 +1,339 @@
+
+        return [label for label in atom_labels if label is not None]
+
+    def _get_frame_atoms_info(self, idx: int) -> dict | None:
```
maybe return an empty dict instead of None so you do not have the ifs

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/trajectory_parsers.py:
```diff
@@ -0,0 +1,339 @@
+        # Cell required
+        if cell is not None:
+            if has_coords('xs', 'ys', 'zs'):
+                positions = self._extract_vector_components(
```
check is positions is not None here

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/trajectory_parsers.py:
```diff
@@ -0,0 +1,339 @@
+                    cell, axis=1
+                )
+            elif has_coords('xsu', 'ysu', 'zsu'):
+                positions = self._extract_vector_components(
```
also here

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/trajectory_parsers.py:
```diff
@@ -0,0 +1,339 @@
+
+        # Unwrapped
+        if positions is None and has_coords('xu', 'yu', 'zu'):
+            positions = self._extract_vector_components(
```
and here

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/trajectory_parsers.py:
```diff
@@ -0,0 +1,339 @@
+
+        # Absolute positions with optional image correction
+        if positions is None and has_coords('x', 'y', 'z'):
+            positions = self._extract_vector_components(frame_atoms_info, 'x', 'y', 'z')
```
and here

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/trajectory_parsers.py:
```diff
@@ -0,0 +1,339 @@
+from nomad_simulation_parsers.parsers.utils.mdanalysisparser import MDAnalysisParser
+
+
+class TrajParser(TextParser):
```
please add short description

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/trajectory_parsers.py:
```diff
@@ -0,0 +1,339 @@
+        ]
+
+
+class TrajParsers:
```
add a short description

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/trajectory_parsers.py:
```diff
@@ -0,0 +1,339 @@
+
+    def __getitem__(self, index: int) -> TrajParser | XYZTrajParser | MDAnalysisParser:
+        if self._parsers is None or not self._parsers:
+            raise IndexError('No parsers available')
```
do not raise error log and return None

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/trajectory_parsers.py:
```diff
@@ -0,0 +1,339 @@
+    def __getitem__(self, index: int) -> TrajParser | XYZTrajParser | MDAnalysisParser:
+        if self._parsers is None or not self._parsers:
+            raise IndexError('No parsers available')
+        return self._parsers[index]
```
you should put a try here instead of the if in l293

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/trajectory_parsers.py:
```diff
@@ -0,0 +1,339 @@
+                    # It's a property/attribute - only return if no args expected
+                    if args or kwargs:
+                        self.logger.warning(
+                            f"Arguments provided for non-callable attribute '{key}'"
```
do not use formatted string

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/trajectory_parsers.py:
```diff
@@ -0,0 +1,339 @@
+
+            except Exception as e:
+                self.logger.debug(
+                    f"Error evaluating '{key}' on {parser.__class__.__name__}: {e}"
```
also here

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/trajectory_parsers.py:
```diff
@@ -0,0 +1,339 @@
+                continue
+
+        if not found_attribute:
+            raise AttributeError(f"Attribute '{key}' not found in parsers")
```
and here

ladinesa commented on src/nomad_simulation_parsers/parsers/utils/constants.py:
maybe simply constants.py?

ladinesa commented on tests/parsers/test_lammps_parser.py:
```diff
@@ -0,0 +1,814 @@
+    assert len(labels) == 320
+
+
+# TODO: re-include output testing alongside migration of parser functionalities
```
remove commented lines we add them later.

ladinesa commented on tests/parsers/test_lammps_parser.py:
```diff
@@ -0,0 +1,814 @@
+            ]
+        return default
+
+    parser.get = mock_get_with_dump
```
you can instead define 
```python 
class TestLogParser(LogParser):
    def parse(self, key):
        if key == 'dump':
            self._results['dump'] = ....
        else:
            super().parse(key)
```

ladinesa commented on tests/parsers/test_lammps_parser.py:
```diff
@@ -0,0 +1,814 @@
+        # Create test data files
+        data_file = os.path.join(tmpdir, 'data.polymer')
+        with open(data_file, 'w') as f:
+            f.write('LAMMPS data file. CGCMM style. atom_style full generated by VMD\n')
```
I do not prefer this file ops. Maybe try:
```
from io import StringIO
f = StringIO()
f.name = 'test'
f.write(....)
f.seek(0)
parser.mainfile = f
```

Bernadette-Mohr replied:
We do a lot of directory scanning, so I don't think we can completely drop creating tempfiles.
But I reduced the number of file operations and limited the tests for `get_data_files` to only focus on `check_file_header` (not also used by `get_traj_files`) using your StringIO suggestion.
Let me know if I am now missing anything.

Bernadette-Mohr committed:
reduced number of helper functions again

Bernadette-Mohr committed:
Minor fixes

Bernadette-Mohr committed:
Changed monkeypatching to using test class in test_lammps_parser

Bernadette-Mohr committed:
Replaced using temporary files for testing file content pattern recognition in get_data_files > check_file_header

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,333 @@
+        )
+        self._data_parser = DataParser()
+
+        self.units = self._log_parser.units
```
not necessary to define

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/parser.py:
```diff
@@ -0,0 +1,333 @@
+        self._parse_content_sections()
+
+        # Close all parser instances
+        self._mdanalysistraj_parser.close()
```
close also log and aux_log parsers

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/file_parsers.py:
```diff
@@ -0,0 +1,799 @@
+
+        return data
+
+    def find_best_matching_file(
```
this looks like a good candidate for a common utils function but could be moved in separate pr. maybe put a TODO

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/trajectory_parsers.py:
```diff
@@ -0,0 +1,362 @@
+                else:
+                    # It's a property/attribute - only return if no args expected
+                    if args or kwargs:
+                        self.logger.warning(
```
remove formatting

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/trajectory_parsers.py:
```diff
@@ -0,0 +1,362 @@
+                    val = parser_attr
+
+            except Exception as e:
+                self.logger.debug(
```
also here

ladinesa commented on src/nomad_simulation_parsers/parsers/lammps/trajectory_parsers.py:
```diff
@@ -0,0 +1,362 @@
+                continue
+
+        if not found_attribute:
+            raise AttributeError("Attribute '%s' not found in parsers", key)
```
do not raise error

ladinesa commented on src/nomad_simulation_parsers/parsers/utils/constants.py:
```diff
@@ -0,0 +1,9 @@
+
+RE_FLOAT = r'[-+]?\d+\.*\d*(?:[Ee][-+]\d+)?'
+RE_N = r'[\n\r]'
+# TODO: Why not use ureg.avogadro_constant or ureg.avogadro_number from pint?
```
Not sure if your units would be an issue when setting your quantity. you can do maybe 1* ureg.avogadros_number.to_base_units(), can you maybe test this?

ladinesa replied:
A quick check, it does work, no issue with setting quantity. you do not need to define this constant here then.
```
from nomad.metainfo import Quantity, MSection
from nomad.units import ureg
import numpy as np
class A(MSection):
    q = Quantity(type=np.float64, unit='joule')

a = A()
a.q = 1*ureg.kJ/ureg.avogadro_number
```

Bernadette-Mohr replied:
Tested with the Lammps parser  `file_parsers.get_unit` function, `(1 * ureg.avogadro_number).to_base_units()` seems to work. We can discuss with @JFRudzinski if we want to replace the hardcoded constant in the parsers at some point.

I also realized that that value we use is outdated, so outsourcing the maintenance of constant definitions might really be something to think about.

ladinesa commented on tests/parsers/test_lammps_parser.py:
```diff
@@ -0,0 +1,586 @@
+    assert os.path.isabs(traj_files[0])
+
+    # Test 2: Scan directory (no dump command)
+    with tempfile.TemporaryDirectory() as tmpdir:
```
maybe make this as a fixture see for instance `nomad-FAIR/tests/conftest.py/tmp`

ladinesa commented:
This looks very good now! just have a couple minor issues you can merge it when resolved.

Bernadette-Mohr committed:
Last minor fixes

Bernadette-Mohr committed:
Fixed error in __init__.py
